torch/csrc/autograd/engine.cpp

```diff
@@ -309,6 +313,21 @@ auto Engine::thread_init(int device) -> void {
   }
 }
 
+// The guard that sets and restores current_graph_task.
+struct GraphTaskGuard {
+  GraphTaskGuard(const std::shared_ptr<GraphTask>& graph_task) {
```

Comment:
You're taking ownership of GraphTask, so you should accept `graph_task` as shared_ptr by value (not const ref)